Store per-difficulty metadata where it differs

Aergwyn opened the pull request:
I went ahead and created a PR to solve #1374. Just close if I was too eager here (regarding possible details I missed because I didn't understand the TODO in the code).

Now it displays the author per beatmap again.

Aergwyn committed:
removed line that set metadata per beatmap to null

peppy commented:
Have you checked the mappings in the database for this? Do they reference the set-level metadata when there's no differences?

Aergwyn commented:
Oh... no, will do later today.

Aergwyn committed:
compare metdata and remove duplicate from beatmap to prevent redundant storage

Aergwyn commented:
Oh. Forgot to mention. It did save for every beatmap even if duplicate (that's why it was set to null previously).
I implemented an Equals (for content equality, at least as far as possible) check and now it only stores metadata if it differs from the beatmapset metadata.

huoyaoyuan commented on osu.Game/Beatmaps/BeatmapMetadata.cs:
```diff
@@ -66,5 +66,23 @@ public string AuthorString
             Tags
         }.Where(s => !string.IsNullOrEmpty(s)).ToArray();
+
+        public override bool Equals(object other)
```
Don't override `object.Equals`. Use `IEqutable<T>`.

Aergwyn replied:
Good catch. Looking for references how this was done before I totally missed this detail. Will change.

huoyaoyuan replied:
I did a lookup for all `object.Equals` references, and the only remained instance was `CultureInfo`.

huoyaoyuan commented on osu.Game/Beatmaps/BeatmapMetadata.cs:
```diff
@@ -66,5 +66,23 @@ public string AuthorString
+            var otherMetadata = other as BeatmapMetadata;
+            if (otherMetadata == null) return false;
+
+            return (onlineBeatmapSetID?.Equals(otherMetadata.onlineBeatmapSetID) ?? false)
```
Use `operator ==` for primitive types.
_Defining custom `==` needs some more work, and seems bothering sometimes._

Aergwyn replied:
Oh right. I will do the same with PreviewTime then.

Aergwyn committed:
requested changes

use IEquatable<T> instead of overriding Equals and `==` operator for primitive types.

huoyaoyuan commented on osu.Game/Beatmaps/BeatmapMetadata.cs:
```diff
@@ -66,5 +67,23 @@ public string AuthorString
+                return false;
+
+            return onlineBeatmapSetID == other.onlineBeatmapSetID
+                && (Title?.Equals(other.Title) ?? false)
```
I meant use `==` for all these lines 😦 
```csharp
=> other != null
&& onlineBeatmapSetID == other.onlineBeatmapSetID
.......
```

Aergwyn replied:
Oh my god. Too much Java. Will fix.

Aergwyn committed:
only use `==` for comparion on primitive types

peppy committed:
Merge branch 'master' into fix-missing-metadata

peppy commented:
We likely want to de-dupe in cases where multiple difficulties in the same set share metadata. Here's a few examples of what I mean:

![sqlitestudio__3_1_1_](https://user-images.githubusercontent.com/191335/33251200-c70b7b04-d378-11e7-9ba1-6c5902c27f3d.png)

![screen shot 2017-11-27 at 13 43 43](https://user-images.githubusercontent.com/191335/33251236-04167436-d379-11e7-9a66-4f6b4eff52f6.png)

Aergwyn commented:
Just preventing duplicates in that regard (I noticed that too, even nekodex on a fresh database sometimes creates 2 entries) or removing duplicates if existing?

peppy commented:
Preventing is fine. We're still early in development so I'm not too concerned about migration/fixing databases when we can just re-import.

Aergwyn committed:
prevent inserting duplicate metadata

Aergwyn committed:
remove unnecessary using

Aergwyn commented:
This is the only location pre insert that I could get working to prevent duplicate entries. I'm not sure if I need to check the database too because if I saw correctly there has to be a BeatmapSet entry for a BeatmapMetadata entry and thus any import would return early anyways.
If it's wrong maybe it works as an inspiration how to do it?

Aergwyn commented:
Everytime when I want to fall asleep I get weird thoughts:
Could it happen that metadata is equal on different beatmapsets? If yes: Does delete cascade?
In a theoretical scenario where both is answered with yes it could reuse metadata entries over multiple beatmapsets and may do some funny things once you delete one of them.

peppy commented:
The de-duplication should *only be between beatmaps in the same set*.

Aergwyn commented:
Yeah, what I thought. I just created the theory that I'd accidentally prune over different sets which I debunked myself (sneaky edits). So this proposed state would fulfill the requirements.

peppy committed:
Merge branch 'master' into fix-missing-metadata

peppy commented on osu.Game/Beatmaps/BeatmapStore.cs:
```diff
@@ -32,6 +32,15 @@ public void Add(BeatmapSetInfo beatmapSet)
         {
             var context = GetContext();
 
+            foreach (var beatmap in beatmapSet.Beatmaps.Where(b => b.Metadata != null))
```
i'm not sure we want to do this. we should be leaving the metadata field null where it inherits from its set, i think (so we can differentiate between set and beatmap metadata when edits occur).

Aergwyn replied:
It is `null` though. The import makes sure of that.
This part just checks for duplicate metadata where metadata is present.
![qq](https://user-images.githubusercontent.com/27222621/33712552-256b71e6-db48-11e7-8071-a508ff18f1c3.png)

peppy replied:
So the `.Local` is checking yet-to-be-committed changes which is why this works? would definitely be worth adding a comment explaining what this is doing, as it's not immediately obvious.

Aergwyn committed:
added comments for local context checking

peppy commented on osu.Game/Beatmaps/BeatmapStore.cs:
```diff
@@ -34,11 +34,12 @@ public void Add(BeatmapSetInfo beatmapSet)
                 var contextMetadata = context.Set<BeatmapMetadata>().Local.SingleOrDefault(e => e.Equals(beatmap.Metadata));
                 if (contextMetadata != null)
-                    beatmap.Metadata = contextMetadata;
+                    beatmap.Metadata = contextMetadata; // reuse existing
```
unnecessary comment

Aergwyn committed:
removed redundant comment

peppy commented on osu.Game/Beatmaps/BeatmapStore.cs:
```diff
@@ -34,11 +34,12 @@ public void Add(BeatmapSetInfo beatmapSet)
 
             foreach (var beatmap in beatmapSet.Beatmaps.Where(b => b.Metadata != null))
             {
+                // check local context for metadata so we can reuse duplicates from the same set
```
i was hoping for explanation on what `.Local` and `.Set` are doing here to achieve this..

Aergwyn committed:
rephrased description of local context checking

peppy committed:
Merge branch 'master' into fix-missing-metadata

peppy commented:
looks good 👍
